README.md
rhcjs
=====
OpenShift Node.js module for interfacing with OpenShift from Node.js. 

## Usage

### Install
    
    # install module
    npm install --save rhcjs
    
### Usage - Basic Auth
  
    var rhc = require('rhcjs)({
      username : 'test@test.com',
      password : 'test123',
      target : 'openshift.redhat.com', // optional - defaults to openshift.redhat.com
      domain : 'test' // optional - if ommitted, app operations won't work
    });

### Usage - Bearer Auth
    
    var rhc = require('rhcjs)({
      username : 'test@test.com',
      token : '1a2b3c', // bearer auth token
      domain : 'test'
    });
    
### Specifying a domain after init
This is useful when you need to do a list domains before deciding which one to use with the rhc client. 


### List Apps
    
    rhc.apps(function(err, res){
      // Res will be an array of apps
    });  
  
### Show App Info
    
    rhc.app.show({ app : 'Some app name' }, function(err, res){
      // Res will be app info
    });  
    
### Create App
    
    rhc.app.create({ 
      name : 'Some app name', 
      cartridges : ['Cartridge(s) to create using']
      empty : true|false, // optional - default false. If true, creates with empty git repo
      fromCode : 'git://somePublicGitRepo' // optional - if set, initialises repo with code pulled from specified public git repo
     }, function(err, res){
      
    });  

### Configure an app
    
    rhc.app.configure({ app : 'Some app name', auto_deploy : true, deployment_branch : 'master' }, function(err, res){
      
    });  
    
### Manage environment variables
    
    // List
    rhc.app.env.list({ app : 'Some app name'}, function(err, envvars){});
      
    // Set  
    rhc.app.env.set({ 
      app : 'Some app name',
      vars : [
        { name : 'envVarName', value : 'envVarValue'}
      ]
    }, function(err, envvars){
      
    });
    
    // Unset
    rhc.app.env.unset({ 
      app : 'Some app name',
      vars : [
        { name : 'someVarName' } // note we don't specify the value here to unset
      ]
    });
    
### Delete App
    
    rhc.app.del({ app : 'Some app name' }, function(err, res){
      
    });

### Add SSH Key to user account
    
    rhc.ssh.add({ name : 'Ssh key name', key : 'SSH key content' }, function(err, res){
      
    });  

### Remove an SSH Key
    
    rhc.ssh.remove({ name : 'Ssh key name' }, function(err, res){
      
    });  

### List cartridges

    rhc.cartridges(function(err, res){
      
    });  

### List domains

    rhc.domains.list(function(err, res){
      
    });  
    
### Create domain

    rhc.domains('domainName', function(err, res){
      
    });
  
### Generate Bearer authentication token (initialise rhcjs using username:password)

    rhc.authorize(function(err, res){
      
    });  

### Add Alias

    rhc.alias.add({id : 'appId', name : 'aliasName'}, function(err, createResult){
      
    });
    
## List Aliases

    rhc.alias.create({id : 'appId' }, function(err, aliasesList){
        
    });
  
## Remove Aliases

    rhc.alias.remove({id : 'appId', name : 'aliasName'}, function(err, removeResult){
        
    });
  

### Sample Errors

#### 401

```
{
  "responseBody": "<!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">\n<html><head>\n<title>401 Authorization Required</title>\n</head><body>\n<h1>Authorization Required</h1>\n<p>This server could not verify that you\nare authorized to access the document\nrequested.  Either you supplied the wrong\ncredentials (e.g., bad password), or your\nbrowser doesn't understand how to supply\nthe credentials required.</p>\n<hr>\n<address>Apache/2.2.15 (Red Hat) Server at openshift.com Port 80</address>\n</body></html>\n",
  "code": 401,
  "error": "Not authenticated (401)",
  "url": "https://invalid:invalid@openshift.com/broker/rest/domain/invalid/applications",
  "method": "post",
  "body": {
    "name": "myfirstapp",
    "cartridges": [
      "diy-1.0"
    ],
    "initial_git_url": "empty"
  }
}
```

#### 422

```
{
  "responseBody": {
    "api_version": 1.3,
    "data": null,
    "messages": [
      {
        "exit_code": 100,
        "field": "name",
        "index": null,
        "severity": "error",
        "text": "The supplied application name 'myfirstapp' already exists"
      }
    ],
    "status": "unprocessable_entity",
    "supported_api_versions": [],
    "type": null,
    "version": "1.3"
  },
  "code": 422,
  "error": "The supplied application name 'myfirstapp' already exists (422)",
  "url": "https://test%40example.com:password@openshift.com/broker/rest/domain/mydomain/applications",
  "method": "post",
  "body": {
    "name": "myfirstapp",
    "cartridges": [
      "diy-0.1"
    ]
  }
}
```

#### 404

```
{
  "responseBody": {
    "api_version": 1.3,
    "data": null,
    "messages": [
      {
        "exit_code": 101,
        "field": null,
        "index": null,
        "severity": "error",
        "text": "Application 'mysecondapp' not found."
      }
    ],
    "status": "not_found",
    "supported_api_versions": [],
    "type": null,
    "version": "1.3"
  },
  "code": 404,
  "error": "Application 'mysecondapp' not found. (404)",
  "url": "https://test%40example.com:password@openshift.com/broker/rest/domain/mydomain/application/mysecondapp?include=cartridges"
}
```
    
## Running Tests
    
    npm test
    

##Or for acceptance tests
    
    export  OPENSHIFT_USERNAME="unamehere"
    export OPENSHIFT_PASSWORD="pwdhere"
    ./node_modules/.bin/mocha -A -u exports --recursive -t 30000 test/accept.js
  


CHANGELOG.md
Release 1.1.1 - 2015-04-13 - Gerard Ryan
------------------------------------------------------
* Add npm shrinkwrap file

Release 1.1.0 - 2015-04-08 - Cian Clarke
------------------------------------------------------
* Add `alias` commands - thanks @alejonext for the contribution!

Release 1.0.4 - 2015-03-10 - David Martin
------------------------------------------------------
* Return a useful error message for all non 2xx responses

Release 1.0.3 - 2015-02-27 - David Martin
------------------------------------------------------
* Add App read by id `rhc.app.read`

Release 1.0.2 - 2015-02-24 - Jason Madigan
------------------------------------------------------
* Return responseBody in error handler

Release 1.0.1 - 2015-02-19 - Cian Clarke
------------------------------------------------------
* Better error handling - return status code on errors

Release 1.0.0 - 2015-02-18 - Cian Clarke
------------------------------------------------------
* Change how domains behave - user must specify a domain to perform Apps API operations.

Release 0.4.0 - 2014-11-06 - Cian Clarke
------------------------------------------------------
* Modify SSH Key Add type parsing to support multiple types

Release 0.3.0 - 2014-11-06 - Cian Clarke
------------------------------------------------------
* Add SSH Key Remove support

Release 0.2.0 - 2014-11-06 - Cian Clarke
------------------------------------------------------
* Add configure command

Release 0.1.0 - 2014-11-04 - Cian Clarke
------------------------------------------------------
* Add ssh list, add commands

Release 0.0.1 - 2014-11-04 - Cian Clarke
------------------------------------------------------
* Initial release of rhcjs - includes apps, app create, app show, app delete commands.


lib/base.js
var request = require('request'),
url = require('url'),
path = require('path');

var HTTP_CODES = {
  400: "The operation could not be completed",
  401: "Not authenticated",
  403: "You are not authorized to perform this operation",
  404: "Not Found",
  409: "The operation could not be completed",
  422: "The operation was not valid",
  500: "The server did not respond correctly. Verify the configured Openshift Service is currently available and accessible",
  503: "The configured Openshift Service is currently unavailable"
};

var parseMessage = function(res, body) {
  // sample res
  // {
  //   "statusCode": 422
  // }
  // 
  // sample body
  // {
  //   "status": "unprocessable_entity",
  //   "messages": [{
  //     "exit_code": 100,
  //     "field": "name",
  //     "index": null,
  //     "severity": "error",
  //     "text": "The supplied application name 'myapp' already exists"
  //   }]
  // }

  var messages = [];
  if (body && body.messages && body.messages instanceof Array && body.messages.length > 0) {
    // parse messages array from body
    body.messages.forEach(function (message) {
      messages.push(message.text);
    });
  } else {
    // use a more generic message based on the status code
    messages.push(HTTP_CODES[res.statusCode] || HTTP_CODES[500]);
  }

  return messages.join('. ') + ' (' + res.statusCode + ')';
};

module.exports = function(params, cb){
  var apiBase = 'broker/rest',
  fullPath = path.join(apiBase, params.url),
  authString = '',
  headers = {
    "User-Agent" : "rhcjs",
    "Content-Type" : "application/json"
  },
  target = this.target.replace(/http(s)?:\/\//, ''),
  fullUrl;
  
  // Yes, request allows basic authentication - but it doesn't send it through many 30* redirects
  if (this.username && this.password){
    authString = encodeURIComponent(this.username) + ':' + encodeURIComponent(this.password) + '@';
  }else if (this.token){
    headers['authorization'] = 'Bearer ' + this.token;
  }else{
    return cb('No auth method provided. Please provide username & password or bearer token');
  }
  
  try{
    fullUrl = url.resolve('https://' + authString + target, fullPath);  
  }catch(err){
    return cb({message : "Error resolving URL string:" , error : err, fullPath : fullPath, target : target });
  }
  
  process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0"; // Avoids DEPTH_ZERO_SELF_SIGNED_CERT error for self-signed certs
  
  request({
    json : true,
    url : fullUrl,
    method : params.method || 'get',
    body : params.body,
    headers : headers
  }, function(err, response, body){
    if (err){
      return cb(err);
    }
    if (response.statusCode.toString()[0]!=='2'){
      var msg = parseMessage(response, body);
      return cb({
        responseBody: body,
        code: response.statusCode,
        error: msg,
        url: fullUrl,
        method: params.method,
        body: params.body
      });
    }
    if (typeof body === 'undefined'){
      // Some requests don't return a body
      return cb(null, {ok : true});
    }
    body = body.data || body;
    return cb(null, body);
  });
};
